src/ntt.cpp

```diff
@@ -151,4 +152,48 @@ std::vector<uint32_t> NTT::mul_vec(const std::vector<uint32_t> &a, const std::ve
 std::vector<uint32_t> NTT::conv(const std::vector<uint32_t> &a, const std::vector<uint32_t> &b) {
     std::vector<uint32_t> c = mul_vec(ntt(a, false), ntt(b, false));
     return ntt(c, true);
-}
\ No newline at end of file
+}
+
+std::vector<uint32_t> NTT::conv_and_reduce(const std::vector<uint32_t> &a, const std::vector<uint32_t> &b, uint32_t r, uint32_t s) {
+    auto call_ntt = [this](const std::vector<uint32_t> &x, bool inverse){ return ntt(x, inverse); };
+    std::future<std::vector<uint32_t>> ntt_a = std::async(std::launch::async, call_ntt, a, false);
+    std::future<std::vector<uint32_t>> ntt_b = std::async(std::launch::async, call_ntt, b, false);
+    auto finish_and_reduce = [this](const std::vector<uint32_t> &x, const std::vector<uint32_t> &y, uint32_t r, uint32_t s){
+        auto out = ntt(mul_vec(x, y), true);
+        for(int32_t i = r-1; i >= 0; --i) {
+            auto red = out[r+i] % 2;
+            out[i] = (out[i] % 2) ^ red;
+            out[s+i] = (out[s+i] % 2) ^ red;
+            out[r+i] = 0;
+        }
+        return out;
+    };
+
+    std::future<std::vector<uint32_t>> conv = std::async(std::launch::async, finish_and_reduce, ntt_a.get(), ntt_b.get(), r, s);
+    return conv.get();
+}
+
+std::pair<std::vector<uint32_t>, std::vector<uint32_t>> NTT::raz_iteration(const std::vector<uint32_t> &product, const std::vector<uint32_t> &delta, uint32_t r, uint32_t s) {
+    auto call_ntt = [this](const std::vector<uint32_t> &x, bool inverse, bool plusone = false){ return ntt(x, inverse, plusone); };
```

Comment:
I'd suggest `std::bind` here as well although you might then lose the default value for the `plusone` parameter which would then require it to be specified explicitly on the below lines, so not sure if that's a fair trade-off?